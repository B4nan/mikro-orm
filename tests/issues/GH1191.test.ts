import {
  Embeddable,
  Embedded,
  Entity,
  Logger,
  MikroORM,
  Platform,
  PrimaryKey,
  Property,
  Type,
} from '@mikro-orm/core';
import { PostgreSqlDriver, PostgreSqlPlatform } from '@mikro-orm/postgresql';

export class Numeric extends Type<number, string> {

  public convertToDatabaseValue(value: number, platform: Platform): string {
    this.validatePlatformSupport(platform);
    return value.toString();
  }

  public convertToJSValue(value: string, platform: Platform): number {
    this.validatePlatformSupport(platform);
    return Number(value);
  }

  public getColumnType(): string {
    return 'numeric(14,2)';
  }

  private validatePlatformSupport(platform: Platform): void {
    if (!(platform instanceof PostgreSqlPlatform)) {
      throw new Error('Numeric custom type implemented only for PG.');
    }
  }

}

@Embeddable()
class Savings {

  @Property({ type: Numeric })
  amount: number;

  constructor(amount: number) {
    this.amount = amount;
  }

}

@Entity()
class User {

  @PrimaryKey()
  id!: number;

  @Embedded()
  savings!: Savings;

  @Property({ nullable: true })
  after?: number; // property after embeddables to verify order props in resulting schema

}

describe('embedded entities in postgresql', () => {
  let orm: MikroORM<PostgreSqlDriver>;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [Savings, User],
      dbName: 'mikro_orm_test_embeddables_custom_types',
      type: 'postgresql',
    });
    await orm.getSchemaGenerator().ensureDatabase();
    await orm.getSchemaGenerator().dropSchema();
    await orm.getSchemaGenerator().createSchema();
  });

  afterAll(() => orm.close(true));

  test('schema', async () => {
    await expect(
      orm.getSchemaGenerator().getCreateSchemaSQL(false)
    ).resolves.toMatchSnapshot('embeddables custom types 1');
    await expect(
      orm.getSchemaGenerator().getUpdateSchemaSQL(false)
    ).resolves.toMatchSnapshot('embeddables custom types 2');
    await expect(
      orm.getSchemaGenerator().getDropSchemaSQL(false)
    ).resolves.toMatchSnapshot('embeddables custom types 3');
  });

  test('persist and load', async () => {
    const user = new User();
    user.savings = new Savings(15200.23);

    const mock = jest.fn();
    const logger = new Logger(mock, ['query']);
    Object.assign(orm.config, { logger });
    await orm.em.persistAndFlush(user);
    orm.em.clear();
    expect(mock.mock.calls[0][0]).toMatch('begin');
    expect(mock.mock.calls[2][0]).toMatch('commit');

    const u = await orm.em.findOneOrFail(User, user.id);
    expect(mock.mock.calls[3][0]).toMatch(
      'select "e0".* from "user" as "e0" where "e0"."id" = $1 limit $2'
    );
    expect(u.savings).toBeInstanceOf(Savings);
    expect(u.savings).toEqual({
      amount: 15200.23,
    });

    await orm.em.flush();
    expect(mock.mock.calls[4][0]).toMatch('begin');
    expect(mock.mock.calls[5][0]).toMatch(
      'update "user" set "savings_amount" = $1 where "id" = $2'
    );
    expect(mock.mock.calls[6][0]).toMatch('commit');
    orm.em.clear();

    const u1 = await orm.em.findOneOrFail(User, {
      savings: { amount: 15200.23 },
    });
    expect(mock.mock.calls[7][0]).toMatch(
      'select "e0".* from "user" as "e0" where "e0"."savings_amount" = $1 limit $2'
    );
    expect(u1.savings.amount).toBe(15200.23);
  });
});
