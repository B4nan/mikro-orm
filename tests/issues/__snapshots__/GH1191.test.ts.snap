// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`embedded entities in postgresql schema: embeddables custom types 1 1`] = `
"create table \\"user\\" (\\"id\\" serial primary key, \\"savings_amount\\" numeric(14,2) not null, \\"after\\" int4 null);

"
`;

exports[`embedded entities in postgresql schema: embeddables custom types 2 1`] = `""`;

exports[`embedded entities in postgresql schema: embeddables custom types 3 1`] = `
"drop table if exists \\"user\\" cascade;

"
`;
